drivers: sensor: mcp9600: Added driver and sample for mcp9600

riedlse opened the pull request:
Added driver and sample for mcp9600 microchip I2C thermocouple
amplifier. Tested on frdm_kw41z and nrf52850dk.

Signed-off-by: Steven Riedl <steve@serconsultingllc.com>

riedlse committed:
drivers: sensor: mcp9600: Added driver and sample for mcp9600

Added driver and sample for mcp9600 microchip I2C thermocouple
amplifier. Tested on frdm_kw41z and nrf52850dk.

Fixed multiple comments on the code. Tested fixes

Signed-off-by: Steven Riedl <steve@serconsultingllc.com>

pabigot commented on drivers/sensor/mcp9600/mcp9600.c:
```diff
@@ -0,0 +1,139 @@
+	if (ret < 0) {
+		return ret;
+	}
+	data->ttemp = (buf[0]<<8) | (buf[1]);
```
I believe spaces around the `<<` operator is more standard, though that might not be present in all existing code.  The array reference operand `buf[1]` doesn't need parentheses around it; the precedence should be clear enough for anybody even if they don't remember whether `<<` or `|` is higher.

riedlse replied:
ok, check patch seems not to care, but I can do that.

pabigot commented on drivers/sensor/mcp9600/mcp9600.c:
```diff
@@ -0,0 +1,139 @@
+
+	switch (chan) {
+	case SENSOR_CHAN_AMBIENT_TEMP:
+			temp = data->ttemp>>4;
```
Indentation here looks wrong: two tabs to statement?

riedlse replied:
will check

pabigot commented on drivers/sensor/mcp9600/mcp9600.h:
```diff
@@ -0,0 +1,107 @@
+struct mcp9600_data {
+	const struct device *i2c_master;
+
+	int16_t ttemp;
```
This ordering wastes memory due to padding between the `int16_t` and `uint32_t` members.  Probably best to group them by whole and fraction.  Or simply store the value in an `int32_t` in some units that cover the sensor range and resolution, e.g. mCel, or what appears to be the native resolution: Cel/16.

Otherwise please document the relationship between the whole and fractional values with respect to sign and units.  I.e. is the true temperature `dtemp + dfrac / 1000000`?

riedlse replied:
Actually am not using the other variables, so they were a waste anyway, removed

pabigot commented on drivers/sensor/mcp9600/mcp9600_trigger.c:
```diff
@@ -0,0 +1,214 @@
+ * Copyright (c) 2020 SER Consulting LLC, Steven Riedl
+ * Copyright (c) 2019 Peter Bigot Consulting, LLC
+ *
+ * !!!! NOT DONE OR TESTED, Converted from mcp9808
```
This does not inspire confidence....

riedlse replied:
Missed that, was actually well tested, even ran some fake value insertion tests to insure the negative temps were correct, will remove...

riedlse replied:
Actually trigger was not tested, just removed all trigger functionality. Can be added back in at a later time properly as the device has 4 triggers.

pabigot commented on samples/sensor/mcp9600/README.rst:
```diff
@@ -0,0 +1,43 @@
+Overview
+********
+
+This sample application periodically reads temperature data from the first
```
Please add a link to manufacturer information about this sensor.

riedlse replied:
ok

riedlse replied:
added

pabigot commented on samples/sensor/mcp9600/boards/nrf52840dongle_nrf52840.overlay:
```diff
@@ -0,0 +1,13 @@
+
+&i2c0 {
+	mcp9600@60 {
+		compatible = "microchip,mcp9600";
```
Need the binding `dts/bindings/sensor/microchip,mcp9600`.

riedlse replied:
need to figure this one out

riedlse replied:
Never knew about that... Found and added

pabigot commented on samples/sensor/mcp9600/src/main.c:
```diff
@@ -0,0 +1,63 @@
+				printf("Temperature: %.4f C\n",
+					sensor_value_to_double(&ttemp));
+#else
+				printk("Temperature: %d\n", ttemp.val1);
```
Include units in the output for non-floating point.

riedlse replied:
ok

riedlse replied:
added

pabigot reviewed (changes requested):
Please also squash the commits and update the description to cover the content of the PR.

riedlse commented on drivers/sensor/mcp9600/mcp9600.c:
```diff
@@ -0,0 +1,139 @@
+#endif /* CONFIG_MCP9600_TRIGGER */
+};
+
+DEVICE_AND_API_INIT(mcp9600, DT_INST_LABEL(0), mcp9600_init,
```
OK

riedlse replied:
fixed

riedlse committed:
drivers: sensor: mcp9600: Added driver and sample for mcp9600

editing error on readme no newline at end

Signed-off-by: Steven Riedl <steve@serconsultingllc.com>

MaureenHelm commented on drivers/sensor/CMakeLists.txt:
```diff
@@ -36,7 +36,7 @@ add_subdirectory_ifdef(CONFIG_ISL29035		isl29035)
 add_subdirectory_ifdef(CONFIG_LIS2DH		lis2dh)
 add_subdirectory_ifdef(CONFIG_LIS2DS12		lis2ds12)
-add_subdirectory_ifdef(CONFIG_LIS2DW12          lis2dw12)
+add_subdirectory_ifdef(CONFIG_LIS2DW12		lis2dw12)
```
Please move unrelated whitespace changes to a separate commit or remove from this PR

riedlse replied:
Will fix on new PR

MaureenHelm commented on drivers/sensor/CMakeLists.txt:
```diff
@@ -67,7 +68,7 @@ add_subdirectory_ifdef(CONFIG_SI7060		si7060)
 add_subdirectory_ifdef(CONFIG_STTS751		stts751)
 add_subdirectory_ifdef(CONFIG_SX9500		sx9500)
-add_subdirectory_ifdef(CONFIG_TH02		    th02)
+add_subdirectory_ifdef(CONFIG_TH02			th02)
```
Another unrelated whitespace change

riedlse replied:
Will fix on split PR

MaureenHelm commented on drivers/sensor/mcp9600/mcp9600.c:
```diff
@@ -0,0 +1,119 @@
+
+#define DT_DRV_COMPAT microchip_mcp9600
+
+#define MCP9600_BUS_I2C DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
```
Does this sensor support SPI too? I don't see this macro being used anywhere

riedlse replied:
I2C only, will check why the macro is still there, probably not needed.

riedlse replied:
done on next PR

MaureenHelm commented on drivers/sensor/mcp9600/mcp9600.c:
```diff
@@ -0,0 +1,119 @@
+
+	switch (chan) {
+	case SENSOR_CHAN_AMBIENT_TEMP:
+		temp = (int32_t)(int16_t)data->ttemp;
```
why the double cast?

riedlse replied:
Peter Bigot pointed that out easy way to get it to be sign extended from an uint16_t to int32_t, thought it was kind of elegant. I'll add a comment

pabigot replied:
That'd be from me: the `(int16_t)` is necessary to convert from the unsigned `uint16_t` because the underlying value is signed.  The outer `(int32_t)` I left in my recommendation because IIRC MISRA doesn't like implicit type conversions.

I'm fine with removing the outer one; absent MISRA I wouldn't have it there myself.

MaureenHelm commented on samples/sensor/mcp9600/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+
+cmake_minimum_required(VERSION 3.13.1)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(lm73c)
```
Rename the project to mcp9600

riedlse replied:
fixed, will do a new PR separate and close this one.

pabigot replied:
Why?

pabigot replied:
Just split the commits locally, and force-push an updated branch:

    git push r.whatever +sensor_mcp9600

riedlse replied:
The separate PR is for the request below to split the driver and sample in to two commits, but my git knowledge may be lacking (use it a lot, but this is my first time adding to a large project like this)

MaureenHelm replied:
See the [Contribution Workflow](https://docs.zephyrproject.org/latest/contribute/index.html#contribution-workflow) for some git tips.

MaureenHelm commented on samples/sensor/mcp9600/README.rst:
```diff
@@ -0,0 +1,49 @@
+.. _mcp9600:
+
+mcp9600 Temperature Sensor
```
Please split the driver and sample application into two commits

riedlse replied:
OK easier for me to close this one and create two new ones.

MaureenHelm commented on drivers/sensor/mcp9600/Kconfig:
```diff
@@ -0,0 +1,47 @@
+	help
+	  Enable driver for MCP9600 i2c thermocouple temperature sensor.
+
+if MCP9600
+
+choice
+	prompt "MCP9600 trigger mode"
+	default MCP9600_TRIGGER_NONE
+
+config MCP9600_TRIGGER_NONE
+	bool "No trigger"
+
+config MCP9600_TRIGGER_GLOBAL_THREAD
+	depends on GPIO
+	select MCP9600_TRIGGER
+	bool "Use global thread"
+
+config MCP9600_TRIGGER_OWN_THREAD
+	depends on GPIO
+	select MCP9600_TRIGGER
+	bool "Use own thread"
+
+endchoice
+
+config MCP9600_TRIGGER
+	bool
+
+config MCP9600_THREAD_STACK_SIZE
+	int "Sensor delayed work thread stack size"
+	depends on MCP9600_TRIGGER_OWN_THREAD
+	default 1024
+
+config MCP9600_THREAD_PRIORITY
+	int "MCP9600 thread priority"
+	depends on MCP9600_TRIGGER_OWN_THREAD
+	default 10
+
+endif # MCP9600
```
This can all be removed since you removed the trigger support

riedlse replied:
Correct, Peter said to move it to dts as well, if I add it back in ever...

MaureenHelm commented on drivers/sensor/mcp9600/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+zephyr_library()
+
+zephyr_library_sources_ifdef(CONFIG_MCP9600 mcp9600.c)
+zephyr_library_sources_ifdef(CONFIG_MCP9600_TRIGGER mcp9600_trigger.c)
```
This line can be removed

riedlse replied:
done

MaureenHelm commented on samples/sensor/mcp9600/sample.yaml:
```diff
@@ -0,0 +1,13 @@
+            - "ttemp: (.*)"
+        fixture: fixture_i2c_mcp9600
+    depends_on: i2c
+    filter: dt_compat_enabled("microchip,mcp9600")
```
I don't think this will pick up the boards you added overlays for, so please replace this with a platform_allow list

riedlse replied:
Switched to the MCP9808 sample.yaml, hope that is acceptable as Peter mentioned not sure what boards/chips are supported.

MaureenHelm reviewed (changes requested):
Thanks for the contribution!

pabigot commented on samples/sensor/mcp9600/boards/nrf52840dongle_nrf52840.overlay:
```diff
@@ -0,0 +1,13 @@
+/*
+ * Copyright (c) 2020, SER Consulting LLC  / Steven Riedl
```
Have you actually tested on this?  It's an unusual board to use for external connections.

If you don't need it, it might be better to drop this and just leave the DK variant of the same SOC.

riedlse replied:
Yes, actually on this (of course you have to wire up an I2C port ) I actually have a custom board that I also use, but makes sense to remove this, mainly tested on the kw and nrfdk.

riedlse replied:
Deleted, will be in next push.

pabigot commented on samples/sensor/mcp9600/README.rst:
```diff
@@ -0,0 +1,49 @@
+        Temperature: 20.8750 C
+        Temperature: 20.6250 C
+
+.. _`MCP9600 Datasheet`: https://ae-bst.resource.bosch.com/media/_tech/media/datasheets/BST-BME280-DS002.pdf
```
No, this is the BME280 data sheet.    Better to link to https://www.microchip.com/wwwproducts/en/MCP9600 which allows people to find whatever they want about this sensor, including but not limited to the datasheet.

riedlse replied:
ARGGGGGG, could have sworn that was in there, trying to be careful with the edits and must have missed that one.

riedlse replied:
next commit, done

pabigot commented on samples/sensor/mcp9600/README.rst:
```diff
@@ -0,0 +1,49 @@
+
+.. code-block:: console
+
+        $ screen /dev/tty.usbmodem0006210000001 115200
```
That seems unlikely to be what anybody else would use.  Maybe `/dev/ttyACM0` like most other examples that suggest using screen?

riedlse replied:
Not sure it makes that much of a difference? I do think I say I tested it on a Mac and that's what you get.

riedlse replied:
added additional text.

pabigot commented on samples/sensor/mcp9600/prj.conf:
```diff
@@ -0,0 +1,9 @@
+CONFIG_SENSOR=y
+CONFIG_MCP9600=y
+CONFIG_BUILD_OUTPUT_HEX=y
+CONFIG_NEWLIB_LIBC=y
```
You shouldn't need newlib now the Zephyr natively knows how to format floating point.  You do need  CBPRINTF_FP_SUPPORT.

riedlse replied:
I will test it without that.

riedlse replied:
done, tested

pabigot commented on samples/sensor/mcp9600/prj.conf:
```diff
@@ -0,0 +1,9 @@
+CONFIG_GPIO=y
+CONFIG_SENSOR=y
+CONFIG_MCP9600=y
+CONFIG_BUILD_OUTPUT_HEX=y
```
Should not need to specify the build output.

riedlse replied:
Actually for the KW41Z the flash seems to prefer BIN although will work with Hex, so I put that in there to eliminate a flash warning. Guess it could disappear.

pabigot commented on samples/sensor/mcp9600/src/main.c:
```diff
@@ -0,0 +1,63 @@
+
+				sensor_channel_get(mcp,
+					SENSOR_CHAN_AMBIENT_TEMP, &ttemp);
+#ifdef CONFIG_NEWLIB_LIBC_FLOAT_PRINTF
```
You're selecting support for floating point output in Kconfig, so you don't need to make it conditional in here.

I'm suspecting there's some samples you used as a template that need to get cleaned up....

riedlse replied:
That would be correct. should be conditional on CBPRINTF_FP_SUPPORT ?

riedlse replied:
CBPRINTF works removed newlib

pabigot commented on drivers/sensor/mcp9600/mcp9600.c:
```diff
@@ -0,0 +1,119 @@
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+
+	ret = mcp9600_reg_read(dev, MCP9600_REG_TEMP_HOT, buf, size);
```
Replace `size` with `sizeof(buf)`.

riedlse replied:
fixed

pabigot commented on drivers/sensor/mcp9600/mcp9600.c:
```diff
@@ -0,0 +1,119 @@
+	if (ret < 0) {
+		return ret;
+	}
+	data->ttemp = (buf[0] << 8) | buf[1];
```
If you make this:

    data->ttemp = (int16_t)((buf[0] << 8) | buf[1]);

and declare `ttemp` as signed you get rid of the cast when the value is extracted, and store the value in format closer to ready-to-use (Cel/16).  You could do `(int16_t)(uint16_t)((buf[0] << 8) | buf[1])` if necessary (shouldn't be, except again for safety-related code guidelines).

riedlse replied:
Never trusted compilers to not sign extend when they shouldn't and to not do it when they should. I prefer any of the more explicit methods. I think there is a micro in zephyr to do the byte swap as well. Would have to go through the other drivers to find it again.

riedlse replied:
actually made it a int32, moved the sign change and multiply up so I could use a value of 1 to invalidate the value on a read failure. There actually is a safety component to my use of this device.

riedlse replied:
tested

pabigot commented on drivers/sensor/mcp9600/mcp9600.c:
```diff
@@ -0,0 +1,119 @@
+	uint8_t buf[2];
+
+	int rc = mcp9600_reg_read(dev, MCP9600_REG_ID_REVISION,
+			buf, 2);
```
s/2/sizeof(buf)/.

riedlse replied:
In this case, yes.

pabigot commented:
Also: squash the second commit which fixes a whitespace issue diagnosed in the first commit.  Unless @MaureenHelm wants the sample in a separate commit from the driver as with #31064; that's not an expectation I have.

Re https://github.com/zephyrproject-rtos/zephyr/pull/31064#discussion_r552972211
> OK, will pull this one and do separate, I guess driver has to go first and be merged?

No, put both commits in the same PR, driver first, then sample.

riedlse committed:
drivers: sensor: mcp9600: Added driver for mcp9600

Split driver and sample in to seperate commits.

This is the entire driver functionality to read the
temperature of a K type thermocouple. Additional code
will be needed if a different type is used or to configure
the alarm trigger functionality.

On branch sensor_mcp9600
Your branch is up to date with 'origin/sensor_mcp9600'.

Changes to be committed:
modified:   drivers/sensor/CMakeLists.txt
modified:   drivers/sensor/Kconfig
modified:   drivers/sensor/mcp9600/CMakeLists.txt
modified:   drivers/sensor/mcp9600/Kconfig
modified:   drivers/sensor/mcp9600/mcp9600.c
modified:   drivers/sensor/mcp9600/mcp9600.h
modified:   dts/bindings/sensor/microchip,mcp9600.yaml

Signed-off-by: Steven Riedl <steve@serconsultingllc.com>

riedlse committed:
drivers: sensor: mcp9600: sample: Added sample for mcp9600

Sample code for using the mcp9600 driver. Sample only
commit.

Signed-off-by: Steven Riedl <steve@serconsultingllc.com>

On branch sensor_mcp9600
Your branch is up to date with 'origin/sensor_mcp9600'.

Changes to be committed:
modified:   samples/sensor/mcp9600/CMakeLists.txt
modified:   samples/sensor/mcp9600/README.rst
deleted:    samples/sensor/mcp9600/boards/nrf52840dongle_nrf52840.overlay
modified:   samples/sensor/mcp9600/prj.conf
modified:   samples/sensor/mcp9600/sample.yaml
modified:   samples/sensor/mcp9600/src/main.c

riedlse committed:
drivers: sensor: mcp9600: Added driver for mcp9600

copy paste error included another driver I am working on that
will be in a later contribution. deleted offending lines.

On branch sensor_mcp9600
Your branch is up to date with 'origin/sensor_mcp9600'.

Changes to be committed:
modified:   drivers/sensor/CMakeLists.txt
modified:   drivers/sensor/Kconfig

Signed-off-by: Steven Riedl <steve@serconsultingllc.com>

riedlse commented:
Don't have time to finish this right now, will probably resubmit later, with proper multi-device support

riedlse commented:
wanted to close on last comment
